Non collaborative close

da-kami opened the pull request:
fixes #73 

Looks can be improved, but should be enough to test if we get till refund.

Ideas that can be done on top of this work:
- [should] add tooltip or similar to all the actions so the user gets more info. 
- [could] the user if he/she know what they do when pressing the commit button. 
- [could] move the commit action button into the dropdown and not show it in actions. Not sure about that.

(would fill them in tickets once this goes in)

Maker screenshot:

![image](https://user-images.githubusercontent.com/5557790/135448857-c647be1e-a1c2-4f4e-9cee-6532ecf9125d.png)

da-kami committed:
Simplify state transition code

The complexity of the code part was not worth the information we got out of it, so it was simplified.
Should hopefully make it a bit more readable and easier to understand.

da-kami committed:
Non-collaborative close by publishing commit tx

da-kami committed:
Re-trigger sending commit in case we start up in `PendingCommit`

For now the behavior is: Once the user pressed the button commit will be published and also retried upon startup.

thomaseizinger commented on daemon/src/routes_taker.rs:
```diff
@@ -96,6 +96,27 @@ pub async fn post_order_request(
+) -> Result<status::Accepted<()>, status::BadRequest<String>> {
+    match action {
+        CfdAction::Accept | CfdAction::Reject => {
+            unreachable!("The taker does not accept and reject");
```
Should just be bad request I think, not a panic.

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -405,41 +405,32 @@ impl Cfd {
-                        bail!("State already assumes lock finality: ignoring")
-                    }
-                    Rejected { .. } | Refunded { .. } | SetupFailed { .. } => {
-                        bail!("The cfd is already in final state {}", self.state)
+                monitor::Event::LockFinality(_) => {
```
Could potentially collapse the two `match` patterns into one.

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -549,6 +579,43 @@ impl Cfd {
     }
 
+    pub fn commit_tx(&self) -> Result<Transaction> {
+        let dlc = if let CfdState::Open { dlc, .. } | CfdState::PendingOpen { dlc, .. } =
```
Could have another accessor or that returns `Option<Dlc>`, then you could just use `context` to bail out of the case where it is `None`.

thomaseizinger commented on daemon/src/routes_taker.rs:
```diff
@@ -96,6 +96,27 @@ pub async fn post_order_request(
+            cfd_actor_address
+                .do_send_async(taker_cfd::Commit { order_id: id })
+                .await
+                .expect("actor to always be available");
```
We should actually also not panic here but return a 500. No network API ever should make it possible to panic an internal process as that can be a security issue.

da-kami replied:
I agree, should fix this in the other API calls at some point, we have several expects there. I only fixed it in the ones concerned by this PR for now, to not cause potential unwanted side effects.

thomaseizinger commented on daemon/src/taker_cfd.rs:
```diff
@@ -85,6 +85,13 @@ impl Actor {
 
+        for cfd in cfds.iter().filter(|cfd| Cfd::is_pending_commit(cfd)) {
+            let signed_commit_tx = cfd.commit_tx()?;
+            let txid = wallet.try_broadcast_transaction(signed_commit_tx).await?;
+
+            tracing::info!("Commit transaction published on chain: {}", txid);
```
Should these be moved into the `Wallet` perhaps?

thomaseizinger reviewed (approved):
Thank you!

Some minor comments bother otherwise good to go in!

klochowicz commented on daemon/src/maker_cfd.rs:
```diff
@@ -19,11 +19,15 @@ use tokio::sync::watch;
 
-pub struct RejectOrder {
+pub struct Reject {
+    pub order_id: OrderId,
```
in the same line as commented on other review, I'll be adding another pair of Accept/Reject structs, so it would be easier if this stayed the same.

da-kami replied:
thanks, missed dropping that one here

klochowicz commented on daemon/src/taker_cfd.rs:
```diff
@@ -301,6 +312,26 @@ impl Actor {
         Ok(())
     }
+
+    // TODO: Duplicated with maker
```
if you'd like to reduce duplication, you could extract a function that would look like this when invoked:

`handle_commit(&mut conn, self.wallet, cfd).await?`

klochowicz commented on daemon/src/taker_cfd.rs:
```diff
@@ -301,6 +312,26 @@ impl Actor {
     }
+
+    // TODO: Duplicated with maker
+    async fn handle_commit(&mut self, msg: Commit) -> Result<()> {
```
in the `handle_***` functions, you don't need to pass the msg, but just `order_id` in this case (like in other handlers)

klochowicz reviewed (approved):
just a bunch of nits :) nice one!

da-kami committed:
Don't panic in endpoints

Does not tackle fixing previously already existing endpoints.

da-kami committed:
Pass order_id instead of msg to handle func

thomaseizinger commented on daemon/src/routes_taker.rs:
```diff
@@ -104,13 +104,14 @@ pub async fn post_cfd_action(
                 .do_send_async(taker_cfd::Commit { order_id: id })
                 .await
-                .expect("actor to always be available");
+                .map_err(|e| status::BadRequest(Some(e.to_string())))?;
```
Nit: I don't think we should leak this error, I will just say "Disconnected".

thomaseizinger reviewed (approved):
Thanks!

I think in this case, `git absorb` would have done a good job here.

da-kami committed:
Integrate settlement endpoint into actions

Caused trouble because of route collisions, so the integration was done now.
Required distinguishing Maker and Taker on the cfd level, which comes in handy anyway.
We distinguish by order origin, but still map to a `Role` enum (`Taker` / `Maker`) for readability.
To avoid duplication the `Role` from contract setup was moved into the model and used throughout.

klochowicz commented on daemon/src/model/cfd.rs:
```diff
@@ -39,12 +39,29 @@ impl<'v> FromParam<'v> for OrderId {
+    Taker,
+}
+
+impl From<Origin> for Role {
```
Perhaps add a comment that this assumption only holds until we only allow the maker to be the only one creating the offers?

da-kami replied:
I think it always holds, because if you are in the role that created the offer, then you are always the maker. 

Sidenote: I think the order is not actually and order. We will have to refactor that eventually. We might have an order at some point, but I don't think it will represent what it currently is.

klochowicz commented on daemon/src/to_sse_event.rs:
```diff
@@ -221,12 +222,15 @@ fn into_unix_secs(time: SystemTime) -> u64 {
-    match state {
-        model::cfd::CfdState::IncomingOrderRequest { .. } => {
+fn actions_for_state(state: model::cfd::CfdState, role: Role) -> Vec<CfdAction> {
+    match (state, role) {
```
nice! I'll use that really soon for accepting/rejecting the settlement.

klochowicz reviewed (approved):
coolness! thanks for reworking the taker settlement into the action :)
